### PR TITLE
任务七: 接入 Tripo / Meshy 闭源 3D API,打通 UE5 / Unity 资产导入链路

### PR DESCRIPTION
**Pull request by DongYu2005**
## 做了什么

**Part A 模型层** 给 `gen_3d_object` 槽位增加两个闭源 API 后端,和现有的 `Trellis2Model` 完全可互换(R6),operator 一行没改就能换后端。

| 文件 | 内容 |
|---|---|
| `agent_skills/develop_harness/api_model_require.md` | 新增契约 R9,闭源 API 专用 |
| `models/common/cloud_api.py` | 共用的 HTTP 重试、错误分类、响应缓存、提交轮询下载 |
| `models/common/glb_utils.py` | 纯标准库的 GLB 解析(面数、材质、包围盒) |
| `models/gen_3d_object/tripo_model.py` | `TripoModel`,Tripo v3 API |
| `models/gen_3d_object/meshy_model.py` | `MeshyModel`,Meshy v1/v2 API |
| `test/test_api_3d_object.py` | 29 个离线契约测试 |

`run.py` 增加 `--backend {trellis2,tripo,meshy}`(默认值不变)以及 `--cache-dir` `--low-poly` `--format` `--timeout` `--decimation-target`。

**Part B 引擎导入** 生成的 mesh 进 UE5 和 Unity。

| 文件 | 运行在哪 |
|---|---|
| `engine_adapters/ue5/import_generated/import_mesh.py` | Unreal Python |
| `engine_adapters/unity3d/import_generated/ImportGeneratedMesh.cs` | Unity Editor |
| `scripts/import_generated_asset.py` | 宿主机,找到编辑器、启动导入器、读回 JSON 报告 |

两边同一套 `--usage {asset,vfx_standalone,vfx_particle}` 分档,默认档不动网格。放在 `import_generated/` 子目录,和任务一要迁的引擎接口代码分开,不会撞。

## 和 model_require.md 的冲突

原契约是照**本地权重模型**写的,闭源 API 有几条天然对不上。这些冲突不是绕过,是新增 `api_model_require.md` 的理由,每一处都标在三个地方:文件头 docstring 的 `CONTRACT DEVIATIONS` 段、冲突那行的 `# [CONTRACT-DEVIATION Cx]` 行内注释、以及新契约里的正式条款。

| id | 原条款 | 为什么对不上 | 怎么处理 |
|---|---|---|---|
| C1 | R2.1 第一个位置参数是权重位置 | API 没有权重文件 | `model_path` 改为承载模型版本号(`"v3.1-20260211"`),**参数名和位置不变**,operator 不需要分支 |
| C2 | R2.3 必须支持 `device="cpu"` | 远程执行,device 没有意义 | 接受但忽略,`cpu` 不报错 |
| C3 | R3.6 推理包 `torch.no_grad()` | 不涉及 torch | 不加装饰器,不 import torch |
| C4 | R4.1 `unload()` 释放显存 | 没有显存 | 实现为关闭 HTTP session,幂等 |
| C5 | R3.3/R3.4 同 seed 同输出 | 服务端不保证可复现 | 支持就透传,不支持就接受后忽略,docstring 写明。**绝不本地伪造可复现性** |
| C6 | R3.1 隐含同步返回 | 异步任务制,提交到完成 5 到 10 分钟 | 内部完成轮询,对调用方仍是同步阻塞;超时异常里带 `task_id`,任务可事后捞回 |
| C7 | R1.2 不构造输出路径 | 产物是 GLB 文件 | 走 R1.2 **本来就写明的例外条款**,路径由调用方传入。列出来只为完整,不算违规 |
| C8 | 契约没覆盖 | 每次调用真花钱 | 新增 `cache_dir`,相同请求零网络返回 |
| C9 | 契约没覆盖 | 网络失败是常态不是异常 | 新增 `max_retries` + 指数退避,可重试(5xx/429/连接失败)与终止(400/401/402)分开,终止错误绝不重试 |

**C8 和 C9 对任务四同样成立**,audio / video 的闭源 API 会遇到一模一样的问题,所以这两块公共逻辑放在 `models/common/cloud_api.py` 而不是每家抄一遍,直接复用。

## 验证情况

离线,无网络无 key 无 GPU:

- `python test/harness/smoke.py` 全 kind 通过,`--kind 3d_object --backend tripo|meshy` 各自通过
- `python test/test_api_3d_object.py` 29/29 通过,覆盖:缺 key fail fast、`device="cpu"`、缓存命中零请求、终止错误不重试、5xx 重试、超时异常带 `task_id`、与 `Trellis2Model` 签名一致

真实服务:

- 两家 API 各生成一次成功,产物落在 `paths.py` 规定的位置,`meta.json` 记录 task_id / 耗时 / credits / 面数 / cache_key
- 缓存验证:相同请求 283 秒变 0.03 秒,余额不变,零重复计费
- 错误分类验证:一次 400(参数越界)和一次 403(余额不足)都被判终止且未重试

引擎(用真实生成的资产):

- UE 5.7:mesh + 材质 + 全部贴图落盘为 `.uasset`,单位换算正确(1 米 → 100uu)
- Unity 6000.5.2f1 + glTFast:prefab,面数与源文件一致,贴图全部绑定
- `--usage vfx_particle` 两个引擎都验过 pivot 归位和尺度归一
- 一条命令 `--engine both --summary` 批量导入两个引擎,2/2 成功

**没做**:FBX / OBJ 路径、多材质资产、UE 侧减面。这几项不在 `model_require.md` 要求内,需要时再补。

## 需要确认的决策

1. **文生 3D 按交接文档的方案 A**:`infer()` / `infer_and_save()` 同时接受 `prompt=` 和 `image=`,给图时以图为准。operator 保持只吃图,`paths.py` 一个字没改。将来要做文生 3D 再加 task kind 即可。
2. **新增了 `models/common/` 目录**。重试、错误分类、缓存、轮询对所有闭源 API 完全一样,放这里共用,写进 R9.11。任务四直接复用。本 PR 唯一的结构新增。
3. **`decimation_target` 会被 clamp 到各家的合法区间**。operator 传的是 TRELLIS.2 的默认值 100 万,两家 API 都会拒。不 clamp 的话调用方就得为每个后端写分支,R6 的可互换性就假了。

## 顺带修的一个 bug(不在任务七范围)

`pipeline/common/paths.py` 写 `meta.json` / `run_meta.json` / summary 时用的是 `write_text()` 不带 encoding,即系统本地编码。中文 Windows 下落盘是 GBK,任何按 UTF-8 读的地方都会炸。仓库路径里带中文就会触发。现在 `paths.py`、`smoke.py` 和两个集成测试的读写全部显式 UTF-8。受影响的只有 `test_data/outputs/` 下的生成产物,已 gitignore,不需要迁移。

## 怎么试

```bash
export MESHY_API_KEY=...      # 或 TRIPO_API_KEY

# 先看额度,balance() 不计费
python -c "from models.gen_3d_object import MeshyModel; print(MeshyModel().balance())"

# 生成,--cache-dir 保证重跑不重复计费
python pipeline/assets_gen/gen_3d_object/run.py --backend meshy \
    --game gameA_cyberpunk_shooter --run-id auto \
    --cache-dir test_data/outputs/_api_cache

# 把这次 run 的产物导进两个引擎
python scripts/import_generated_asset.py --engine both \
    --summary test_data/outputs/<game>/<run>/3d_object_results_summary.json
```

没有 key 也能验:`python test/harness/smoke.py --kind 3d_object --backend tripo`

一次调用大约 30 到 40 credits、5 到 10 分钟。**跑导入前先关掉引擎工程**,导入器会自己起编辑器进程,开着的那个看不到新资产。

契约条款见 `agent_skills/develop_harness/api_model_require.md`,引擎侧使用注意见 `engine_adapters/{ue5,unity3d}/import_generated/README.md`。

## 兼容性

默认后端仍是 `trellis2`,不带新 flag 时行为和改动前完全一致;operator 返回的 dict key 没有删改;legacy flat 模式仍在 `smoke.py` 里验。

### CODE REVIEW

**Comment by zengbohan0217** [comment]
有一点点建议微调的地方，
1. 检查一下 agent_skills/develop_harness/ 这个下面新改的三个文件，尽量别让AI把 代码修改记录提交上来
2. 我在考虑 类似于这些闭源API skill，要不要单独维护一个py文件，或者其他文件也行，列一个借口列表，方便AI Agent查看，这里可以和AI对一下
3. requirements.txt 之前没写，是因为各个环境比较难统一，目前都是放到了 scripts/installing 这个下面，不过环境配置也是skill，尤其是几个引擎的，需要让AI加一下
4. engine_adapters/unity3d/import_generated/ 以及 engine_adapters/ue5/import_generated/  下面的内容我暂时没检查，可以和一帆确认下，一些接口可以同步下，最后汇总到 agent_skills/engine_context 下面
